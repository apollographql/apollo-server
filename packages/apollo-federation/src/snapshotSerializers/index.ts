export { default as astSerializer } from './astSerializer';
export { default as selectionSetSerializer } from './selectionSetSerializer';
export { default as typeSerializer } from './typeSerializer';
export { default as graphqlErrorSerializer } from './graphqlErrorSerializer';

declare global {
  namespace jest {
    interface Expect {
      /**
       * Adds a module to format application-specific data structures for serialization.
       */
      addSnapshotSerializer(serializer: import('pretty-format').Plugin): void;
    }
  }
}
