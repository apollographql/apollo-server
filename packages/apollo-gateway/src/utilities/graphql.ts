import {
  ASTNode,
  FieldNode,
  GraphQLCompositeType,
  GraphQLField,
  GraphQLInterfaceType,
  GraphQLNullableType,
  GraphQLObjectType,
  GraphQLSchema,
  GraphQLType,
  GraphQLUnionType,
  isListType,
  isNonNullType,
  Kind,
  ListTypeNode,
  NamedTypeNode,
  OperationDefinitionNode,
  parse,
  print,
  SchemaMetaFieldDef,
  SelectionNode,
  TypeMetaFieldDef,
  TypeNameMetaFieldDef,
  TypeNode,
} from 'graphql';

/**
 * Not exactly the same as the executor's definition of getFieldDef, in this
 * statically evaluated environment we do not always have an Object type,
 * and need to handle Interface and Union types.
 */
export function getFieldDef(
  schema: GraphQLSchema,
  parentType: GraphQLCompositeType,
  fieldName: string,
): GraphQLField<any, any> | undefined {
  if (
    fieldName === SchemaMetaFieldDef.name &&
    schema.getQueryType() === parentType
  ) {
    return SchemaMetaFieldDef;
  }
  if (
    fieldName === TypeMetaFieldDef.name &&
    schema.getQueryType() === parentType
  ) {
    return TypeMetaFieldDef;
  }
  if (
    fieldName === TypeNameMetaFieldDef.name &&
    (parentType instanceof GraphQLObjectType ||
      parentType instanceof GraphQLInterfaceType ||
      parentType instanceof GraphQLUnionType)
  ) {
    return TypeNameMetaFieldDef;
  }
  if (
    parentType instanceof GraphQLObjectType ||
    parentType instanceof GraphQLInterfaceType
  ) {
    return parentType.getFields()[fieldName];
  }

  return undefined;
}

export function getResponseName(node: FieldNode): string {
  return node.alias ? node.alias.value : node.name.value;
}

export function allNodesAreOfSameKind<T extends ASTNode>(
  firstNode: T,
  remainingNodes: ASTNode[],
): remainingNodes is T[] {
  return !remainingNodes.some(node => node.kind !== firstNode.kind);
}

export function astFromType(
  type: GraphQLNullableType,
): NamedTypeNode | ListTypeNode;
export function astFromType(type: GraphQLType): TypeNode {
  if (isListType(type)) {
    return { kind: Kind.LIST_TYPE, type: astFromType(type.ofType) };
  } else if (isNonNullType(type)) {
    return { kind: Kind.NON_NULL_TYPE, type: astFromType(type.ofType) };
  } else {
    return {
      kind: Kind.NAMED_TYPE,
      name: { kind: Kind.NAME, value: type.name },
    };
  }
}

export function printWithReducedWhitespace(ast: ASTNode): string {
  return print(ast)
    .replace(/\s+/g, ' ')
    .trim();
}

export function parseSelections(source: string): ReadonlyArray<SelectionNode> {
  return (parse(`query { ${source} }`)
    .definitions[0] as OperationDefinitionNode).selectionSet.selections;
}
