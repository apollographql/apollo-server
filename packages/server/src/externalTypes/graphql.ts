/**
 * This file exports types related specifically to GraphQL execution, most
 * notably `GraphQLRequest` and `GraphQLResponse`, used most heavily by
 * `executeHTTPGraphQLRequest` and `executeOperation`. The
 * `responseForOperation` plugin hook also returns a `GraphQLResponse`.
 */
import type { FormattedExecutionResult } from 'graphql';
import type { BaseContext } from './context.js';
import type { HTTPGraphQLHead, HTTPGraphQLRequest } from './http.js';
import type { WithRequired } from '@apollo/utils.withrequired';
import type {
  GraphQLExperimentalFormattedInitialIncrementalExecutionResult,
  GraphQLExperimentalFormattedSubsequentIncrementalExecutionResult,
} from './incrementalDeliveryPolyfill.js';

export interface GraphQLRequest<
  TVariables extends VariableValues = VariableValues,
> {
  query?: string;
  operationName?: string;
  variables?: TVariables;
  extensions?: Record<string, any>;
  http?: HTTPGraphQLRequest;
}

export type VariableValues = { [name: string]: any };

// A GraphQL response can either be a single result, or an initial result
// followed by a stream of subsequent results. The latter occurs when the
// GraphQL operation uses incremental delivery directives such as `@defer` or
// `@stream`. Note that incremental delivery currently requires using a
// pre-release of graphql-js v17.
export type GraphQLResponseBody<TData = Record<string, unknown>> =
  | {
      kind: 'single';
      singleResult: FormattedExecutionResult<TData>;
    }
  | {
      kind: 'incremental';
      initialResult: GraphQLExperimentalFormattedInitialIncrementalExecutionResult;
      subsequentResults: AsyncIterable<GraphQLExperimentalFormattedSubsequentIncrementalExecutionResult>;
    };

export type GraphQLInProgressResponse<TData = Record<string, unknown>> = {
  http: HTTPGraphQLHead;
  body?: GraphQLResponseBody<TData>;
};

export type GraphQLResponse<TData = Record<string, unknown>> = WithRequired<
  GraphQLInProgressResponse<TData>,
  'body'
>;
export interface ExecuteOperationOptions<TContext extends BaseContext> {
  contextValue?: TContext;
}
