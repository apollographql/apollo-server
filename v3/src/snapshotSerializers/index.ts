export { default as astSerializer } from './astSerializer';

declare global {
  namespace jest {
    interface Expect {
      /**
       * Adds a module to format application-specific data structures for serialization.
       */
      addSnapshotSerializer(serializer: import('pretty-format').Plugin): void;
    }
  }
}
