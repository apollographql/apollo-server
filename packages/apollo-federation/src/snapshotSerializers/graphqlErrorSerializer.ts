import { GraphQLError } from 'graphql';
import { Plugin } from 'pretty-format';

export default {
  test(value: any) {
    return value && value instanceof GraphQLError;
  },

  print(value: GraphQLError, print) {
    return print({
      message: value.message,
      code: value.extensions ? value.extensions.code : 'MISSING_ERROR',
    });
  },
} as Plugin;
