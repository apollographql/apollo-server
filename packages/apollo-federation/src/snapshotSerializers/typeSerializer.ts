import { isNamedType, GraphQLNamedType, printType } from 'graphql';
import { Plugin } from 'pretty-format';

export default {
  test(value: any) {
    return value && isNamedType(value);
  },
  print(value: GraphQLNamedType) {
    return printType(value);
  },
} as Plugin;
