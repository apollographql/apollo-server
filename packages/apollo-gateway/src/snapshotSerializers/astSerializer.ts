import { ASTNode, print } from 'graphql';
import { Plugin, Config, Refs } from 'pretty-format';

export default {
  test(value: any) {
    return value && typeof value.kind === 'string';
  },

  serialize(
    value: ASTNode,
    _config: Config,
    indentation: string,
    _depth: number,
    _refs: Refs,
    _printer: any,
  ): string {
    return print(value)
      .trim()
      .replace(/\n/g, '\n' + indentation);
  },
} as Plugin;
