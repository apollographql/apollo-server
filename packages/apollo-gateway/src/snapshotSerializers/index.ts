import astSerializer from './astSerializer';
import selectionSetSerializer from './selectionSetSerializer';
import typeSerializer from './typeSerializer';
import queryPlanSerializer from './queryPlanSerializer';
export {
  astSerializer,
  selectionSetSerializer,
  typeSerializer,
  queryPlanSerializer,
};

declare global {
  namespace jest {
    interface Expect {
      /**
       * Adds a module to format application-specific data structures for serialization.
       */
      addSnapshotSerializer(serializer: import('pretty-format').Plugin): void;
    }
  }
}
