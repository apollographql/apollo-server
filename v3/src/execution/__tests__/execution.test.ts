import {
  parseGraphqlRequest,
  validateGraphqlRequest,
  executeGraphqlRequest,
  processGraphqlRequest
} from "..";
import { astSerializer } from "../../snapshotSerializers";
import { VariableValues, gql } from "apollo-server-core";
import { buildServiceDefinition } from "@apollographql/apollo-tools";
import { GraphQLError } from "graphql";

expect.addSnapshotSerializer(astSerializer);

// This is a handy trick for syntax highlighting GraphQL query strings in editors
// like VS Code that depend on the gql`` or graphql`` tagged template literals.
const graphql = String.raw;

// An array of tuples, each tuple representing a test.
// A single tuple has the shape: [testName, { ...testParams }]
type JestInput<T> = Record<string, T>;

const validQueries: JestInput<{
  query: string;
  operationName?: string;
  variables?: VariableValues;
}> = {
  "basic - { people { name } }": {
    query: graphql`
      query Basic {
        people {
          name
        }
      }
    `
  },
  "with nested selection sets - { people { name friends { name friends { name } } } }": {
    query: graphql`
      query Nested {
        people {
          name
          friends {
            name
            friends {
              name
            }
          }
        }
      }
    `
  },
  "with directives - { people { name { friends { name } } }": {
    query: graphql`
      query Directives($includeFriends: Boolean!) {
        people {
          name
          friends @include(if: $includeFriends) {
            name
          }
        }
      }
    `,
    variables: { includeFriends: true }
  },
  "with multiple documents - { people { name } }": {
    query: graphql`
      query FirstDocument {
        people {
          name
        }
      }

      query SecondDocument {
        people {
          id
        }
      }
    `,
    operationName: "FirstDocument"
  }
};

// Using graphql`` on these unparseable document breaks code highlighting and other
// tooling dependent on graphql's parser.
const unparseableQueries: JestInput<string> = {
  "missing closing }": `
      query Basic {
        people {
          name
        }
    `,
  "empty selection set": `
      query Nested {
        people { }
      }
    `,
  "empty query": `
      query Empty {}
    `
};

const invalidQueries: JestInput<{ query: string; operationName?: string }> = {
  basic: {
    query: graphql`
      query Basic {
        invalidField
      }
    `
  },
  "with invalid selection sets": {
    query: graphql`
      query Nested {
        people {
          name
          invalidField
        }
      }
    `
  },
  "missing directive": {
    query: graphql`
      query MissingDirective {
        people @invalid {
          name
        }
      }
    `
  },
  "with multiple errors": {
    query: graphql`
      query FirstDocument {
        people {
          invalidOne
        }
        invalidTwo
      }
    `
  }
};

const nonExecutableQueries: JestInput<{ query: string }> = {
  "error field throws": {
    query: graphql`
      query {
        error
      }
    `
  },
  "throws multiple errors": {
    query: graphql`
      query {
        error
        people {
          nestedError
        }
      }
    `
  }
};

const serverTeam = [
  { id: 1, name: "James Baxley" },
  { id: 2, name: "Ashi Krishnan" },
  { id: 3, name: "Jesse Rosenberger" },
  { id: 4, name: "Trevor Scheer" }
];

const serviceDefinition = buildServiceDefinition([
  {
    typeDefs: gql`
      type Query {
        people: [Person]
        error: String
        modifiesContext: String
      }

      type Person {
        id: ID
        name: String
        friends: [Person]
        nestedError: String
      }
    `,
    resolvers: {
      Query: {
        people() {
          return serverTeam;
        },
        error() {
          throw new GraphQLError("Error while resolving `error` field");
        },
        modifiesContext(_, __, context) {
          context.modified = true;
          return "Context modified!";
        }
      },
      Person: {
        async friends(parentValue) {
          return serverTeam.filter(({ id }) => id !== parentValue.id);
        },
        nestedError() {
          throw new GraphQLError("Error while resolving `nestedError` field");
        }
      }
    }
  }
]);

const schema = serviceDefinition.schema!;

const validQueryTests = Object.entries(validQueries);
const unparseableQueryTests = Object.entries(unparseableQueries);
const invalidQueryTests = Object.entries(invalidQueries);
const nonExecutableQueryTests = Object.entries(nonExecutableQueries);

describe("parseGraphqlRequest", () => {
  it.each(validQueryTests)("Parses valid queries - %s", (_, { query }) => {
    const parseResult = parseGraphqlRequest({ query });
    if ("error" in parseResult) {
      throw new Error(
        "Unexpected parse failure in parseable query. Are you sure you added a valid query to the validQueries test cases?"
      );
    }

    expect(parseResult.document).toMatchSnapshot();
  });

  it.each(unparseableQueryTests)(
    "Returns an error for invalid queries - %s",
    (_, query) => {
      const parseResult = parseGraphqlRequest({ query });
      if ("document" in parseResult) {
        throw new Error(
          "Unexpected successful parse in unparseable query. Are you sure you added an unparseable query to the unparseableQueries test cases?"
        );
      }

      expect(parseResult.error).toMatchSnapshot();
    }
  );
});

describe("validateGraphqlRequest", () => {
  it.each(validQueryTests)(
    "valid queries against schema: %s",
    (_, { query, operationName }) => {
      const parseResult = parseGraphqlRequest({ query });
      if ("error" in parseResult) {
        throw new Error(
          "Unexpected parse failure in parseable query. Are you sure you added a valid query to the validQueries test cases?"
        );
      }

      const validationErrors = validateGraphqlRequest({
        schema,
        document: parseResult.document,
        operationName
      });

      expect(validationErrors).toHaveLength(0);
    }
  );

  it.each(invalidQueryTests)(
    "invalid queries against schema: %s",
    (_, { query, operationName }) => {
      const parseResult = parseGraphqlRequest({ query });
      if ("error" in parseResult) {
        throw new Error(
          "Unexpected parse failure in parseable query. Are you sure you added a parseable query to the invalidQueries test cases?"
        );
      }

      const validationErrors = validateGraphqlRequest({
        schema,
        document: parseResult.document,
        operationName
      });

      expect(validationErrors).toMatchSnapshot();
    }
  );
});

describe("executeGraphqlRequest", () => {
  it.each(validQueryTests)(
    "Executable queries - %s",
    async (_, { query, operationName, variables }) => {
      const parseResult = parseGraphqlRequest({ query });
      if ("error" in parseResult) {
        throw new Error(
          "Unexpected parse failure in parseable query. Are you sure you added a parseable query to the validQueries test cases?"
        );
      }

      const result = await executeGraphqlRequest({
        schema,
        document: parseResult.document,
        operationName,
        variables
      });
      expect(result).toMatchSnapshot();
    }
  );

  it.each(nonExecutableQueryTests)(
    "Errors during execution - %s",
    async (_, { query }) => {
      const parseResult = parseGraphqlRequest({ query });
      if ("error" in parseResult) {
        throw new Error(
          "Unexpected parse failure in parseable query. Are you sure you added a parseable query to the nonExecutableQueries test cases?"
        );
      }

      const { errors } = await executeGraphqlRequest({
        schema,
        document: parseResult.document
      });

      expect(errors!.length).toBeGreaterThan(0);
      expect(errors).toMatchSnapshot();
    }
  );
});

describe("processGraphqlRequest", () => {
  it.each(validQueryTests)(
    "Executable queries - %s",
    async (_, { query, operationName, variables }) => {
      const { data, errors } = await processGraphqlRequest({
        schema,
        request: {
          query,
          operationName,
          variables
        }
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot();
    }
  );

  it.each(unparseableQueryTests)(
    "Unparseable queries - %s",
    async (_, query) => {
      const { errors } = await processGraphqlRequest({
        schema,
        request: {
          query
        }
      });

      expect(errors).toMatchSnapshot();
    }
  );

  it.each(invalidQueryTests)(
    "Invalid queries - %s",
    async (_, { query, operationName }) => {
      const { errors } = await processGraphqlRequest({
        schema,
        request: {
          query,
          operationName
        }
      });

      expect(errors).toMatchSnapshot();
    }
  );

  it.each(nonExecutableQueryTests)(
    "Non-executable queries - %s",
    async (_, { query }) => {
      const { errors } = await processGraphqlRequest({
        schema,
        request: {
          query
        }
      });

      expect(errors!.length).toBeGreaterThan(0);
      expect(errors).toMatchSnapshot();
    }
  );

  it("Passes a modifiable context object to resolvers", async () => {
    const context = Object.create(null);
    const { data } = await processGraphqlRequest({
      schema,
      request: {
        query: graphql`
          query ModifiesContext {
            modifiesContext
          }
        `
      },
      context
    });

    expect(data!.modifiesContext).toBe("Context modified!");
    expect(context).toHaveProperty("modified");
  });
});
