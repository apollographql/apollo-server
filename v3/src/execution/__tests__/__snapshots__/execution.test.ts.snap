// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`executeGraphqlRequest Errors during execution - error field throws 1`] = `
Array [
  [GraphQLError: Error while resolving \`error\` field],
]
`;

exports[`executeGraphqlRequest Errors during execution - throws multiple errors 1`] = `
Array [
  [GraphQLError: Error while resolving \`error\` field],
  [GraphQLError: Error while resolving \`nestedError\` field],
  [GraphQLError: Error while resolving \`nestedError\` field],
  [GraphQLError: Error while resolving \`nestedError\` field],
  [GraphQLError: Error while resolving \`nestedError\` field],
]
`;

exports[`executeGraphqlRequest Executable queries - basic - { people { name } } 1`] = `
Object {
  "data": Object {
    "people": Array [
      Object {
        "name": "James Baxley",
      },
      Object {
        "name": "Ashi Krishnan",
      },
      Object {
        "name": "Jesse Rosenberger",
      },
      Object {
        "name": "Trevor Scheer",
      },
    ],
  },
}
`;

exports[`executeGraphqlRequest Executable queries - with directives - { people { name { friends { name } } } 1`] = `
Object {
  "data": Object {
    "people": Array [
      Object {
        "friends": Array [
          Object {
            "name": "Ashi Krishnan",
          },
          Object {
            "name": "Jesse Rosenberger",
          },
          Object {
            "name": "Trevor Scheer",
          },
        ],
        "name": "James Baxley",
      },
      Object {
        "friends": Array [
          Object {
            "name": "James Baxley",
          },
          Object {
            "name": "Jesse Rosenberger",
          },
          Object {
            "name": "Trevor Scheer",
          },
        ],
        "name": "Ashi Krishnan",
      },
      Object {
        "friends": Array [
          Object {
            "name": "James Baxley",
          },
          Object {
            "name": "Ashi Krishnan",
          },
          Object {
            "name": "Trevor Scheer",
          },
        ],
        "name": "Jesse Rosenberger",
      },
      Object {
        "friends": Array [
          Object {
            "name": "James Baxley",
          },
          Object {
            "name": "Ashi Krishnan",
          },
          Object {
            "name": "Jesse Rosenberger",
          },
        ],
        "name": "Trevor Scheer",
      },
    ],
  },
}
`;

exports[`executeGraphqlRequest Executable queries - with multiple documents - { people { name } } 1`] = `
Object {
  "data": Object {
    "people": Array [
      Object {
        "name": "James Baxley",
      },
      Object {
        "name": "Ashi Krishnan",
      },
      Object {
        "name": "Jesse Rosenberger",
      },
      Object {
        "name": "Trevor Scheer",
      },
    ],
  },
}
`;

exports[`executeGraphqlRequest Executable queries - with nested selection sets - { people { name friends { name friends { name } } } } 1`] = `
Object {
  "data": Object {
    "people": Array [
      Object {
        "friends": Array [
          Object {
            "friends": Array [
              Object {
                "name": "James Baxley",
              },
              Object {
                "name": "Jesse Rosenberger",
              },
              Object {
                "name": "Trevor Scheer",
              },
            ],
            "name": "Ashi Krishnan",
          },
          Object {
            "friends": Array [
              Object {
                "name": "James Baxley",
              },
              Object {
                "name": "Ashi Krishnan",
              },
              Object {
                "name": "Trevor Scheer",
              },
            ],
            "name": "Jesse Rosenberger",
          },
          Object {
            "friends": Array [
              Object {
                "name": "James Baxley",
              },
              Object {
                "name": "Ashi Krishnan",
              },
              Object {
                "name": "Jesse Rosenberger",
              },
            ],
            "name": "Trevor Scheer",
          },
        ],
        "name": "James Baxley",
      },
      Object {
        "friends": Array [
          Object {
            "friends": Array [
              Object {
                "name": "Ashi Krishnan",
              },
              Object {
                "name": "Jesse Rosenberger",
              },
              Object {
                "name": "Trevor Scheer",
              },
            ],
            "name": "James Baxley",
          },
          Object {
            "friends": Array [
              Object {
                "name": "James Baxley",
              },
              Object {
                "name": "Ashi Krishnan",
              },
              Object {
                "name": "Trevor Scheer",
              },
            ],
            "name": "Jesse Rosenberger",
          },
          Object {
            "friends": Array [
              Object {
                "name": "James Baxley",
              },
              Object {
                "name": "Ashi Krishnan",
              },
              Object {
                "name": "Jesse Rosenberger",
              },
            ],
            "name": "Trevor Scheer",
          },
        ],
        "name": "Ashi Krishnan",
      },
      Object {
        "friends": Array [
          Object {
            "friends": Array [
              Object {
                "name": "Ashi Krishnan",
              },
              Object {
                "name": "Jesse Rosenberger",
              },
              Object {
                "name": "Trevor Scheer",
              },
            ],
            "name": "James Baxley",
          },
          Object {
            "friends": Array [
              Object {
                "name": "James Baxley",
              },
              Object {
                "name": "Jesse Rosenberger",
              },
              Object {
                "name": "Trevor Scheer",
              },
            ],
            "name": "Ashi Krishnan",
          },
          Object {
            "friends": Array [
              Object {
                "name": "James Baxley",
              },
              Object {
                "name": "Ashi Krishnan",
              },
              Object {
                "name": "Jesse Rosenberger",
              },
            ],
            "name": "Trevor Scheer",
          },
        ],
        "name": "Jesse Rosenberger",
      },
      Object {
        "friends": Array [
          Object {
            "friends": Array [
              Object {
                "name": "Ashi Krishnan",
              },
              Object {
                "name": "Jesse Rosenberger",
              },
              Object {
                "name": "Trevor Scheer",
              },
            ],
            "name": "James Baxley",
          },
          Object {
            "friends": Array [
              Object {
                "name": "James Baxley",
              },
              Object {
                "name": "Jesse Rosenberger",
              },
              Object {
                "name": "Trevor Scheer",
              },
            ],
            "name": "Ashi Krishnan",
          },
          Object {
            "friends": Array [
              Object {
                "name": "James Baxley",
              },
              Object {
                "name": "Ashi Krishnan",
              },
              Object {
                "name": "Trevor Scheer",
              },
            ],
            "name": "Jesse Rosenberger",
          },
        ],
        "name": "Trevor Scheer",
      },
    ],
  },
}
`;

exports[`parseGraphqlRequest Parses valid queries - basic - { people { name } } 1`] = `
query Basic {
  people {
    name
  }
}
`;

exports[`parseGraphqlRequest Parses valid queries - with directives - { people { name { friends { name } } } 1`] = `
query Directives($includeFriends: Boolean!) {
  people {
    name
    friends @include(if: $includeFriends) {
      name
    }
  }
}
`;

exports[`parseGraphqlRequest Parses valid queries - with multiple documents - { people { name } } 1`] = `
query FirstDocument {
  people {
    name
  }
}

query SecondDocument {
  people {
    id
  }
}
`;

exports[`parseGraphqlRequest Parses valid queries - with nested selection sets - { people { name friends { name friends { name } } } } 1`] = `
query Nested {
  people {
    name
    friends {
      name
      friends {
        name
      }
    }
  }
}
`;

exports[`parseGraphqlRequest Returns an error for invalid queries - empty query 1`] = `[GraphQLError: Syntax Error: Expected Name, found "}".]`;

exports[`parseGraphqlRequest Returns an error for invalid queries - empty selection set 1`] = `[GraphQLError: Syntax Error: Expected Name, found "}".]`;

exports[`parseGraphqlRequest Returns an error for invalid queries - missing closing } 1`] = `[GraphQLError: Syntax Error: Expected Name, found <EOF>.]`;

exports[`processGraphqlRequest Executable queries - basic - { people { name } } 1`] = `
Object {
  "people": Array [
    Object {
      "name": "James Baxley",
    },
    Object {
      "name": "Ashi Krishnan",
    },
    Object {
      "name": "Jesse Rosenberger",
    },
    Object {
      "name": "Trevor Scheer",
    },
  ],
}
`;

exports[`processGraphqlRequest Executable queries - with directives - { people { name { friends { name } } } 1`] = `
Object {
  "people": Array [
    Object {
      "friends": Array [
        Object {
          "name": "Ashi Krishnan",
        },
        Object {
          "name": "Jesse Rosenberger",
        },
        Object {
          "name": "Trevor Scheer",
        },
      ],
      "name": "James Baxley",
    },
    Object {
      "friends": Array [
        Object {
          "name": "James Baxley",
        },
        Object {
          "name": "Jesse Rosenberger",
        },
        Object {
          "name": "Trevor Scheer",
        },
      ],
      "name": "Ashi Krishnan",
    },
    Object {
      "friends": Array [
        Object {
          "name": "James Baxley",
        },
        Object {
          "name": "Ashi Krishnan",
        },
        Object {
          "name": "Trevor Scheer",
        },
      ],
      "name": "Jesse Rosenberger",
    },
    Object {
      "friends": Array [
        Object {
          "name": "James Baxley",
        },
        Object {
          "name": "Ashi Krishnan",
        },
        Object {
          "name": "Jesse Rosenberger",
        },
      ],
      "name": "Trevor Scheer",
    },
  ],
}
`;

exports[`processGraphqlRequest Executable queries - with multiple documents - { people { name } } 1`] = `
Object {
  "people": Array [
    Object {
      "name": "James Baxley",
    },
    Object {
      "name": "Ashi Krishnan",
    },
    Object {
      "name": "Jesse Rosenberger",
    },
    Object {
      "name": "Trevor Scheer",
    },
  ],
}
`;

exports[`processGraphqlRequest Executable queries - with nested selection sets - { people { name friends { name friends { name } } } } 1`] = `
Object {
  "people": Array [
    Object {
      "friends": Array [
        Object {
          "friends": Array [
            Object {
              "name": "James Baxley",
            },
            Object {
              "name": "Jesse Rosenberger",
            },
            Object {
              "name": "Trevor Scheer",
            },
          ],
          "name": "Ashi Krishnan",
        },
        Object {
          "friends": Array [
            Object {
              "name": "James Baxley",
            },
            Object {
              "name": "Ashi Krishnan",
            },
            Object {
              "name": "Trevor Scheer",
            },
          ],
          "name": "Jesse Rosenberger",
        },
        Object {
          "friends": Array [
            Object {
              "name": "James Baxley",
            },
            Object {
              "name": "Ashi Krishnan",
            },
            Object {
              "name": "Jesse Rosenberger",
            },
          ],
          "name": "Trevor Scheer",
        },
      ],
      "name": "James Baxley",
    },
    Object {
      "friends": Array [
        Object {
          "friends": Array [
            Object {
              "name": "Ashi Krishnan",
            },
            Object {
              "name": "Jesse Rosenberger",
            },
            Object {
              "name": "Trevor Scheer",
            },
          ],
          "name": "James Baxley",
        },
        Object {
          "friends": Array [
            Object {
              "name": "James Baxley",
            },
            Object {
              "name": "Ashi Krishnan",
            },
            Object {
              "name": "Trevor Scheer",
            },
          ],
          "name": "Jesse Rosenberger",
        },
        Object {
          "friends": Array [
            Object {
              "name": "James Baxley",
            },
            Object {
              "name": "Ashi Krishnan",
            },
            Object {
              "name": "Jesse Rosenberger",
            },
          ],
          "name": "Trevor Scheer",
        },
      ],
      "name": "Ashi Krishnan",
    },
    Object {
      "friends": Array [
        Object {
          "friends": Array [
            Object {
              "name": "Ashi Krishnan",
            },
            Object {
              "name": "Jesse Rosenberger",
            },
            Object {
              "name": "Trevor Scheer",
            },
          ],
          "name": "James Baxley",
        },
        Object {
          "friends": Array [
            Object {
              "name": "James Baxley",
            },
            Object {
              "name": "Jesse Rosenberger",
            },
            Object {
              "name": "Trevor Scheer",
            },
          ],
          "name": "Ashi Krishnan",
        },
        Object {
          "friends": Array [
            Object {
              "name": "James Baxley",
            },
            Object {
              "name": "Ashi Krishnan",
            },
            Object {
              "name": "Jesse Rosenberger",
            },
          ],
          "name": "Trevor Scheer",
        },
      ],
      "name": "Jesse Rosenberger",
    },
    Object {
      "friends": Array [
        Object {
          "friends": Array [
            Object {
              "name": "Ashi Krishnan",
            },
            Object {
              "name": "Jesse Rosenberger",
            },
            Object {
              "name": "Trevor Scheer",
            },
          ],
          "name": "James Baxley",
        },
        Object {
          "friends": Array [
            Object {
              "name": "James Baxley",
            },
            Object {
              "name": "Jesse Rosenberger",
            },
            Object {
              "name": "Trevor Scheer",
            },
          ],
          "name": "Ashi Krishnan",
        },
        Object {
          "friends": Array [
            Object {
              "name": "James Baxley",
            },
            Object {
              "name": "Ashi Krishnan",
            },
            Object {
              "name": "Trevor Scheer",
            },
          ],
          "name": "Jesse Rosenberger",
        },
      ],
      "name": "Trevor Scheer",
    },
  ],
}
`;

exports[`processGraphqlRequest Invalid queries - basic 1`] = `
Array [
  [GraphQLError: Cannot query field "invalidField" on type "Query".],
]
`;

exports[`processGraphqlRequest Invalid queries - missing directive 1`] = `
Array [
  [GraphQLError: Unknown directive "@invalid".],
]
`;

exports[`processGraphqlRequest Invalid queries - with invalid selection sets 1`] = `
Array [
  [GraphQLError: Cannot query field "invalidField" on type "Person".],
]
`;

exports[`processGraphqlRequest Invalid queries - with multiple errors 1`] = `
Array [
  [GraphQLError: Cannot query field "invalidOne" on type "Person".],
  [GraphQLError: Cannot query field "invalidTwo" on type "Query".],
]
`;

exports[`processGraphqlRequest Non-executable queries - error field throws 1`] = `
Array [
  [GraphQLError: Error while resolving \`error\` field],
]
`;

exports[`processGraphqlRequest Non-executable queries - throws multiple errors 1`] = `
Array [
  [GraphQLError: Error while resolving \`error\` field],
  [GraphQLError: Error while resolving \`nestedError\` field],
  [GraphQLError: Error while resolving \`nestedError\` field],
  [GraphQLError: Error while resolving \`nestedError\` field],
  [GraphQLError: Error while resolving \`nestedError\` field],
]
`;

exports[`processGraphqlRequest Unparseable queries - empty query 1`] = `
Array [
  [GraphQLError: Syntax Error: Expected Name, found "}".],
]
`;

exports[`processGraphqlRequest Unparseable queries - empty selection set 1`] = `
Array [
  [GraphQLError: Syntax Error: Expected Name, found "}".],
]
`;

exports[`processGraphqlRequest Unparseable queries - missing closing } 1`] = `
Array [
  [GraphQLError: Syntax Error: Expected Name, found <EOF>.],
]
`;

exports[`validateGraphqlRequest invalid queries against schema: basic 1`] = `
Array [
  [GraphQLError: Cannot query field "invalidField" on type "Query".],
]
`;

exports[`validateGraphqlRequest invalid queries against schema: missing directive 1`] = `
Array [
  [GraphQLError: Unknown directive "@invalid".],
]
`;

exports[`validateGraphqlRequest invalid queries against schema: with invalid selection sets 1`] = `
Array [
  [GraphQLError: Cannot query field "invalidField" on type "Person".],
]
`;

exports[`validateGraphqlRequest invalid queries against schema: with multiple errors 1`] = `
Array [
  [GraphQLError: Cannot query field "invalidOne" on type "Person".],
  [GraphQLError: Cannot query field "invalidTwo" on type "Query".],
]
`;
